sasmodels/kernelcl.py

```diff
@@ -317,14 +309,54 @@ def compile_program(self, name, source, dtype, fast, timestamp):
         if program_timestamp < timestamp:
             del self.compiled[key]
         if key not in self.compiled:
-            context = self.get_context(dtype)
+            context = self.context[dtype]
             logging.info("building %s for OpenCL %s", key,
                          context.devices[0].name.strip())
-            program = compile_model(self.get_context(dtype),
+            program = compile_model(self.context[dtype],
                                     str(source), dtype, fast)
             self.compiled[key] = (program, timestamp)
         return program
 
+    def free_buffer(self, key):
+        if key in self.cache:
+            self.cache[key].release()
+            del self.cache[key]
+
+    def __del__(self):
+        for v in self.cache.values():
+            release = getattr(v, 'release', lambda: None)
+            release()
+        self.cache = {}
+
+_CURRENT_ID = 0
+def unique_id():
+    global _CURRENT_ID
+    _CURRENT_ID += 1
+    return _CURRENT_ID
+
+def _create_some_context():
+    # type: () -> cl.Context
+    """
+    Protected call to cl.create_some_context without interactivity.
+
+    Uses SAS_OPENCL or PYOPENCL_CTX if they are set in the environment,
+    otherwise scans for the most appropriate device using
+    :func:`_get_default_context`
+    """
+    if 'SAS_OPENCL' in os.environ:
```

Comment:
Selecting "No OpenCL" deletes the SAS_OPENCL entry from the os.environ dictionary. Switching from an OpenCL option to "No OpenCL" with a fit loaded still results in the 'No fitter' error. The PYOPENCL_CTX environment variable will likely need to get removed whenever the SAS_OPENCL variable does.

Reply:
No OpenCL should be setting SAS_OPENCL=none, then it won't even try to use OpenCL.  Not sure if this will fix the FitDriver problem.

Reply:
With current master and SAS_OPENCL changing from an OpenCL device to no opencl device, the following appears when compute is selected for an existing model:
```
14:17:32 - ERROR : sas.sasgui.perspectives.fitting.fitting:1890: Traceback (most recent call last):
  File "/Users/pkienzle/Source/sasview/src/sas/sascalc/data_util/calcthread.py", line 274, in _run
    self.compute(*args, **kwargs)
  File "/Users/pkienzle/Source/sasview/src/sas/sasgui/perspectives/fitting/model_thread.py", line 198, in compute
    output[index] = self.model.evalDistribution(self.data.x[index])
  File "/Users/pkienzle/Source/sasmodels/sasmodels/sasview_model.py", line 616, in evalDistribution
    return self.calculate_Iq(qdist)
  File "/Users/pkienzle/Source/sasmodels/sasmodels/sasview_model.py", line 678, in calculate_Iq
    return self._calculate_Iq(qx, qy)
  File "/Users/pkienzle/Source/sasmodels/sasmodels/sasview_model.py", line 701, in _calculate_Iq
    magnetic=is_magnetic)
  File "/Users/pkienzle/Source/sasmodels/sasmodels/kernelcl.py", line 610, in __call__
    env = environment()
  File "/Users/pkienzle/Source/sasmodels/sasmodels/kernelcl.py", line 157, in environment
    raise RuntimeError("SAS_OPENCL=None in environment")
RuntimeError: SAS_OPENCL=None in environment
```
It is not enough just to clear the cache of OpenCL models when the context is reset, we may also need to build the kernel on a different kind of device.

So, rather than trying to support a live context switch as this patch does, we should instead make sure that sasview has no live kernels after the context switch.  This can be done in sasview_model.py, which already manages a list of all models.

I will back this patch out of the beta branch will be a bit challenging.  Although it does no harm being in there, the code can be simpler if the kernels assume there will be no context switch.

I will put a new PR in against master which fixes the problem in sasview_model.